File: 카카오/보석 쇼핑.js
function solution(gems) {
    let answer = [0, gems.length - 1];
    
    let gemType = new Set(gems);
    let gemSize = gemType.size;
    let gemMap = new Map();    
    
    let gl = gems.length;
    let start = 0, end = 0;
    gemMap.set(gems[0], 1);
    
    while(end < gl && start <= end) {
        if (gemSize === gemMap.size) {
            if (answer[1] - answer[0] > end - start)
                answer = [start, end];
            if ((answer[1] - answer[0] == end - start) && (start < answer[0]))
                answer = [start, end];
            
            if (gemMap.get(gems[start]) > 1)
                gemMap.set(gems[start], gemMap.get(gems[start]) - 1);
            else
                gemMap.delete(gems[start]);
            start++;
        } else {
            end++;
            gemMap.set(gems[end], 1 + (gemMap.get(gems[end]) || 0));
        }
    }
    
    
    return [answer[0] + 1, answer[1] + 1];
}

File: 프로그래머스/카카오/보석 쇼핑.py
def solution(gems):
    n = len(gems)
    gem_type = len(set(gems))
    
    l, r = 0, 0
    gem_counter = {}
    minval = n
    ans = [1, n]

    while l < n:
        if r < n and len(gem_counter) < gem_type:
            if gems[r] not in gem_counter:
                gem_counter[gems[r]] = 1
            else:
                gem_counter[gems[r]] += 1
            r += 1
        else:
            if len(gem_counter) == gem_type and r - l < minval:
                minval = r - l
                ans = [l + 1, r]

            gem_counter[gems[l]] -= 1
            if gem_counter[gems[l]] == 0:
                del gem_counter[gems[l]]
            l += 1
    
    return ans